Handle error message on transaction deletion

mbayopanda opened the pull request:
This PR is about the issue #4233

mbayopanda committed:
Handle error message on transaction deletion

jniles commented on client/src/modules/journal/modals/editTransaction.modal.js:
```diff
@@ -342,7 +344,8 @@ function JournalEditTransactionController(
             Modal.close(deleteTransactionResult);
-          });
+          })
+          .catch(Notify.handleError);
```
Can you just log into a `<span class="text-danger"> SOMETHING BROKE!!!</span>`?

mbayopanda replied:
I have created a customised one : 

![image](https://user-images.githubusercontent.com/5445251/79020956-355d0c00-7b72-11ea-93aa-028a4290ffa2.png)

mbayopanda committed:
Use of custom component for notification

mbayopanda committed:
remove console reference

mbayopanda committed:
Remove not necessary translate

jniles commented on client/src/js/components/bhModalNotify/bhModalNotify.js:
```diff
@@ -0,0 +1,60 @@
+
+function ModalNotifyController($timeout) {
+  const $ctrl = this;
+  const TTL = $ctrl.ttl || 3000;
```
This should be in an `$onInit()` function.

jniles commented on client/src/js/components/bhModalNotify/bhModalNotify.html:
```diff
@@ -0,0 +1,21 @@
+  }
+</style>
+
+<div ng-if="$ctrl.visible" class="bh-modal-notify {{$ctrl.style}}">
```
It's more the angularjs way to use `ngClass`.  Could you do that here?

mbayopanda replied:
I was thinking about that but ngClass allow value in this format : 
```html
ng-class="{'the_css_class': javascript_variable }"
```

Here in the `bhModalNotify` is not known in advance

jniles commented on client/src/js/components/bhModalNotify/bhModalNotify.html:
```diff
@@ -0,0 +1,21 @@
+<style>
+  .bh-modal-notify {
```
I can tell you've been writing VueJS.  Can you just put this in our bhima-bootstrap stylesheet?

jniles commented on client/src/js/components/bhModalNotify/bhModalNotify.js:
```diff
@@ -0,0 +1,60 @@
+  }
+
+  $ctrl.close = () => {
+    $ctrl.visible = false;
```
You could also cancel the `$timeout` here just to clean some things up.

jniles replied:
This is good.  Well done!

jniles reviewed (commented):
I've tested this and it works.  I'm a little torn on introducing a new UI pattern.  Do you think it is a good idea @mbayopanda?

I've made some clean up comments, mostly to make this more AngularJS-like.  Would it be possible to write some basic unit tests for this component?

mbayopanda committed:
use angularjs style

mbayopanda committed:
remove tag

jniles commented on client/src/js/components/bhModalNotify/bhModalNotify.js:
```diff
@@ -0,0 +1,62 @@
+    controller  : ModalNotifyController,
+    transclude  : true,
+    bindings    : {
+      value : '<?',
```
What does `value` do?

jniles reviewed (commented):
@mbayopanda I don't see any client unit tests at the moment.

The scenarios I would test are:
  1. Ensure that no error message is present if the `error` is undefined.
  2. Ensure that an error message is present when the `error` is defined.
  3. Ensure that the error message is removed when the "x" is clicked.
  4. Ensure that the error message is removed with the TTL expires.

jniles commented on client/src/js/components/bhModalNotify/bhModalNotify.js:
```diff
@@ -0,0 +1,62 @@
+  .component('bhModalNotify', {
+    templateUrl : 'js/components/bhModalNotify/bhModalNotify.html',
+    controller  : ModalNotifyController,
+    transclude  : true,
```
Do you need transclude?  It doesn't look like you actually use it.

jniles committed:
test(bhModalNotify): unit tests for functionality

This commit tests the bhModalNotify component to ensure that it appears
and disappears when it should.

jniles reviewed (commented):
@mbayopanda thanks for working with me on the unit tests.

As discussed, please look at the `type` binding.  I don't think it is used.  Similarly, `$ctrl.status` isn't defined.  I think you might have gotten confused and they should be the same thing.

Also, please remove the `transclude` if it is unused.

If you can think of any more unit tests to write, go for it!

mbayopanda committed:
adding some unit test for the bh-modal-notify

jniles commented on karma.conf.js:
```diff
@@ -18,6 +18,7 @@ module.exports = (config) => {
       'test/client-unit/mocks/*.js',
       `bin/client/js/bhima/bhima.min.js`,
       'bin/client/modules/**/*.html',
+      'bin/client/js/components/**/*.html',
```
👍

jniles commented on test/client-unit/components/bhModalNotify.spec.js:
```diff
@@ -0,0 +1,181 @@
+
+  let $flushPendingTasks;
+
+  const find = (elm, selector) => elm[0].querySelector(selector);
```
Since we've used this in a number of other unit tests, we could probably make a shared utility file to put these in.  I can't remember if we are allowed to `require()` in this context or not though.

jniles commented on test/client-unit/components/bhModalNotify.spec.js:
```diff
@@ -0,0 +1,181 @@
+    const translated = $translate.instant($scope.someValue);
+    $scope.$digest();
+
+    const notificationClass = find(compiled, '.notification-danger');
```
These tests are great!  I think they are fine as they are.  One potential problem is that we may never find the `.notification-danger` class if a bug occurs where danger becomes success.  A way to test for this would be to instead search for the `.notification` class, then expect it to also have `.notification-danger`, like we do for the `fa-times` class.

```js
const notification  = find(compiled, '.notification');
expect(angular.element(notification).hasClass('notification-danger')).to.equal(true);
```

jniles reviewed (approved):
Looks good to me!  Nice additional tests :)

bors r+
